File: src/main/java/gui/Toolbar.java
package gui;

import graph.algorithm.GraphPath;
import graph.edge.*;
import graph.vertex.Vertex;
import network.Network;
import network.TestRunner;
import utils.PropertiesHandler;
import utils.WeightList;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionListener;
import java.util.function.BiConsumer;
import java.util.stream.Collectors;

public class Toolbar extends JPanel {
    private final NetworkPanel networkPanel;
    private final Network network;
    private final JComboBox<Vertex> fromNode;
    private final JComboBox<Vertex> toNode;
    private final JTextArea infoArea;

    public Toolbar(NetworkPanel panel) {
        super();
        this.networkPanel = panel;
        network = panel.getNetwork();
        setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
        Vertex[] nodes = network.getNodes(false).toArray(new Vertex[]{});
        fromNode = new JComboBox<>(nodes);
        toNode = new JComboBox<>(nodes);
        panel1();
        panel2();
        panel3();
        panel4();
        infoArea = new JTextArea();
        JScrollPane jScrollPane = new JScrollPane(infoArea);
        jScrollPane.getViewport().setPreferredSize(new Dimension(400, 800));
        add(jScrollPane);
    }

    private void panel4() {
        JPanel panel4 = new JPanel();
        panel4.setLayout(new BoxLayout(panel4, BoxLayout.X_AXIS));
        panel4.add(createButton("Налаштування", e -> new Settings()));
        add(panel4);
    }

    private void panel3() {
        JPanel panel = new JPanel();
        panel.setLayout(new BoxLayout(panel, BoxLayout.X_AXIS));
        panel.add(createButton("Зв'язати", nodeAction(network::createConnection)));
        panel.add(createButton("Роз'єднання", nodeAction(network::closeConnection)));
        panel.add(createButton("Роз'єднати всі", e -> {
            network.closeAll();
            networkPanel.update();
        }));
        panel.add(createButton("Таблиця", e -> {
            Vertex node = (Vertex) fromNode.getSelectedItem();
            infoArea.setText(network.getPaths(node).stream()
                    .map(GraphPath::toString)
                    .collect(Collectors.joining("\n")));
        }));
        add(panel);
    }

    private void panel2() {
        JPanel pane = new JPanel();
        pane.setLayout(new BoxLayout(pane, BoxLayout.X_AXIS));
        pane.add(createButton("Термінал", nodeAction((n1, n2) -> {
            n1.setTerminal(!n1.isTerminal());
            updateNodes();
        })));
        pane.add(createButton("Додати", e -> {
            network.addNode();
            updateNodes();
            networkPanel.update();
        }));
        pane.add(createButton("Видалити", nodeAction((node, n) -> {
            network.removeNode(node);
            updateNodes();
        })));
        pane.add(createButton("Створити канал", nodeAction((n1, n2) -> {
            PropertiesHandler cfg = PropertiesHandler.getProps();
            int weight = cfg.<WeightList>getProperty("weights").getWeight();
            cfg.<ChannelFactory>getProperty("channelFactory").createChannel(n1, n2, weight);
        })));
        pane.add(createButton("Роз'єднати канал", nodeAction(network::removeConnection)));
        add(pane);
    }

    private void panel1() {
        JPanel panel = new JPanel();
        panel.setLayout(new BoxLayout(panel, BoxLayout.X_AXIS));
        panel.add(fromNode);
        panel.add(toNode);
        panel.add(createButton("Видалити все", e -> {
            network.clear();
            updateNodes();
            networkPanel.update();
        }));
        panel.add(createButton("Згенерувати", e -> {
            PropertiesHandler.getProps().setProperty("counter", 0);
            network.generateNetwork();
            updateNodes();
            networkPanel.update();
        }));
        panel.add(createButton("Запустити тести", e -> {
            infoArea.setText("");
            network.closeAll();
            infoArea.setText(TestRunner.getConfigOptions() +
                    "\nВіртуальний канал\n" +
                    new TestRunner(network).runTests(false) +
                    "\nДейтаграмний режим\n" +
                    new TestRunner(network).runTests(true));
        }));
        add(panel);
    }

    private Component createButton(String name, ActionListener action) {
        JButton button = new JButton(name);
        button.addActionListener(action);
        return button;
    }

    private void updateNodes() {
        fromNode.removeAllItems();
        toNode.removeAllItems();
        network.getNodes(false).forEach(node -> {
            fromNode.addItem(node);
            toNode.addItem(node);
        });
    }

    private ActionListener nodeAction(BiConsumer<Vertex, Vertex> consumer) {
        return e -> {
            Vertex from = (Vertex) fromNode.getSelectedItem();
            Vertex to = (Vertex) toNode.getSelectedItem();
            consumer.accept(from, to);
            networkPanel.update();
        };
    }
}

File: src/main/java/mvc/view/Toolbar.kt
package mvc.view

import mvc.model.channel.ChannelFactory
import mvc.model.node.Node
import mvc.controller.Network
import common.Config
import common.WeightList
import mvc.controller.Informator

import javax.swing.*
import java.awt.*

class Toolbar(private val networkPanel: NetworkPanel) : JPanel() {
    private val network: Network = networkPanel.network
    private val fromNode: JComboBox<Node>
    private val toNode: JComboBox<Node>
    private val timeLabel: JLabel
    private val infoArea: JTextArea

    init {
        layout = BoxLayout(this, BoxLayout.Y_AXIS)
        val nodes = network.getNodes().toTypedArray()
        fromNode = JComboBox(nodes)
        toNode = JComboBox(nodes)
        addNodePanel()
        addChangePanel()
        addInfoPanel()
        timeLabel = JLabel("0")
        infoArea = Config.config.getProperty("info")
        infoArea.isEditable = false
        add(infoArea)
    }

    private fun addInfoPanel() {
        val panel = JPanel()
        panel.layout = BoxLayout(panel, BoxLayout.X_AXIS)
        panel.add(createButton("Видалити всі вузли") {
            Config.config.setProperty("counter", 0)
            network.clear()
            updateNodes()
            networkPanel.update()
        })
        panel.add(createButton("Нова мережа") {
            Config.config.setProperty("counter", 0)
            network.generateNetwork()
            updateNodes()
            networkPanel.update()
        })
        panel.add(createButton("Табл.маршрутизації") {
            val node = fromNode.selectedItem as Node
            infoArea.text = network.getPaths(node)
                    .joinToString("\n") { it.toString() }
            Config.config.getProperty<JTabbedPane>("tab").selectedIndex = 2
        })
        panel.add(createButton("Серія тестів") {
            infoArea.text = ""
            network.closeAll()
            infoArea.text = Informator.configOptions +
                    Informator(network).runTests()
            Config.config.getProperty<JTabbedPane>("tab").selectedIndex = 2
        })
        add(panel)
    }

    private fun addChangePanel() {
        val pane = JPanel()
        pane.layout = BoxLayout(pane, BoxLayout.X_AXIS)
        pane.add(createButton("Роб. станція", nodeAction { n1, _ ->
            n1.isTerminal = !n1.isTerminal
            updateNodes()
        }))
        pane.add(createButton("Обрати", nodeAction { from, to ->
            val current = from.isSelected
            if (current == to.isSelected) {
                val changed = !current
                network.getPath(from, to).forEach { ch -> ch.isSelected = changed }
                from.isSelected = changed
                to.isSelected = changed
            }
        }))
        pane.add(createButton("Новий вузол") {
            network.addNode()
            updateNodes()
            networkPanel.update()
        })
        pane.add(createButton("Видалити вузол", nodeAction { node, _ ->
            network.removeNode(node)
            updateNodes()
        }))
        pane.add(createButton("Додати канал", nodeAction { n1, n2 ->
            val cfg = Config.config
            val weight = cfg.getProperty<WeightList>("weights").weight
            cfg.getProperty<ChannelFactory>("channelFactory").createChannel(n1, n2, weight)
        }))
        pane.add(createButton("Видалити канал", nodeAction { from, to ->
            network.removeConnection(from, to) }))
        add(pane)
    }

    private fun addNodePanel() {
        val panel = JPanel()
        panel.layout = BoxLayout(panel, BoxLayout.X_AXIS)
        panel.add(fromNode)
        panel.add(toNode)
        panel.add(createButton("З'єднання", nodeAction { from, to ->
            network.createConnection(from, to)
        }))
        panel.add(createButton("Роз'єднання", nodeAction { from, to ->
            network.closeConnection(from, to)
        }))
//        panel.add(createButton("Роз'єднати все") {
//            mvc.controller.closeAll()
//            networkPanel.update()
//        })
        add(panel)
    }

    private fun createButton(name: String, action: () -> Unit): Component {
        val button = JButton(name)
        button.addActionListener { action() }
        return button
    }

    private fun updateNodes() {
        fromNode.removeAllItems()
        toNode.removeAllItems()
        network.getNodes().forEach {
            fromNode.addItem(it)
            toNode.addItem(it)
        }
    }

    private fun nodeAction(consumer: (Node, Node) -> Unit): () -> Unit {
        return {
            val from = fromNode.selectedItem as Node
            val to = toNode.selectedItem as Node
            consumer(from, to)
            networkPanel.update()
        }
    }
}
